autonomic-administration-plugin/src/test/java/br/com/autonomiccs/autonomic/administration/plugin/AdministrationAgentTest.java

```diff
@@ -0,0 +1,665 @@
+        setupMapAndExecuteVMsMigrationsTest(hosts, true);
+
+        administrationAgent.mapAndExecuteVMsMigrations(cluster, clusterAdministrationHeuristicAlgorithm);
+
+        verifyMapAndExecuteVMsMigrationsCalledMethods(0, 0);
+    }
+
+    @Test
+    public void mapAndExecuteVMsMigrationsTestOneHost() {
+        ClusterVO cluster = createCluster(null);
+        List<HostResources> hosts = createHosts(1);
+        setupMapAndExecuteVMsMigrationsTest(hosts, true);
+
+        administrationAgent.mapAndExecuteVMsMigrations(cluster, clusterAdministrationHeuristicAlgorithm);
+
+        verifyMapAndExecuteVMsMigrationsCalledMethods(0, 0);
+    }
+
+    @Test
+    public void mapAndExecuteVMsMigrationsTestFullExecution() {
+        ClusterVO cluster = createCluster(null);
+        List<HostResources> hosts = createHosts(2);
+        setupMapAndExecuteVMsMigrationsTest(hosts, true);
+
+        administrationAgent.mapAndExecuteVMsMigrations(cluster, clusterAdministrationHeuristicAlgorithm);
+
+        verifyMapAndExecuteVMsMigrationsCalledMethods(1, 1);
+    }
+
+    @Test
+    public void mapAndExecuteVMsMigrationsTestEmptyMigrationsMap() {
+        ClusterVO cluster = createCluster(null);
+        List<HostResources> hosts = createHosts(2);
+        setupMapAndExecuteVMsMigrationsTest(hosts, false);
+
+        administrationAgent.mapAndExecuteVMsMigrations(cluster, clusterAdministrationHeuristicAlgorithm);
+
+        verifyMapAndExecuteVMsMigrationsCalledMethods(1, 0);
+    }
+
+    @Test
+    public void migrateVMTestTargetHostEqualsVmHost() throws Exception {
+        VMInstanceVO vmInstanceVo = migrateVMSetupVmInstanceVo(1l);
+        HostVO hostVo = migrateVMSetupHostVo(1l);
+        Mockito.doReturn(vmInstanceVo).when(userVmService).migrateVirtualMachine(Mockito.anyLong(), Mockito.any(HostVO.class));
+        Mockito.doReturn(new HostVO("")).when(hostService).findHostById(Mockito.anyLong());
+
+        administrationAgent.migrateVM(vmInstanceVo, hostVo);
+
+        Mockito.verify(userVmService, Mockito.times(0)).migrateVirtualMachine(Mockito.anyLong(), Mockito.any(HostVO.class));
+    }
+
+    @Test
+    public void migrateVMTestTargetHostDifferentFromVmHost() throws Exception {
+        VMInstanceVO vmInstanceVo = migrateVMSetupVmInstanceVo(0l);
+        HostVO hostVo = migrateVMSetupHostVo(1l);
+        Mockito.doReturn(vmInstanceVo).when(userVmService).migrateVirtualMachine(Mockito.anyLong(), Mockito.any(HostVO.class));
+
+        administrationAgent.migrateVM(vmInstanceVo, hostVo);
+
+        Mockito.verify(userVmService, Mockito.times(1)).migrateVirtualMachine(Mockito.anyLong(), Mockito.any(HostVO.class));
+    }
+
+    @Test
+    public void shutdownHostTestHostWithoutVms() {
+        setupShutdownHostTest(new ArrayList<VMInstanceVO>());
+        administrationAgent.shutdownHost(0l);
+        verifyShutdownHostCalledMethods(1);
+    }
+
+    @Test
+    public void shutdownHostTestHostWithVms() {
```

Comment:
you are missing the flow of execution when an exception happens here, it has to be tested too